modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,494 @@
+                }
+
+                // Prepare direct buffer slice to read keys from the iterator. It's goind to be used to read row ids without timestamps.
+                // To enforce this, we limit it to a proper capacity.
+                ByteBuffer directBuffer = MV_KEY_BUFFER.get()
+                        .position(0).limit(ROW_ID_OFFSET + igniteRowIdSize).slice().order(BIG_ENDIAN);
+
+                while (true) {
+                    // This flag is used to skip row ids that were created after required timestamp.
+                    boolean found = true;
+
+                    // We should do it all the time. Here in particular it means one of two things:
+                    //  - partition is empty;
+                    //  - iterator exhausted all the data in partition.
+                    if (invalid(it)) {
+                        return false;
+                    }
+
+                    // At this point, seekKeyBuf should contain row id that's above the one we already scanned, but not greater than any
+                    // other row id in partition. When we start, row id is filled with zeroes. Value during the iteration is described later
+                    // in this code. Now let's descrie what we'll find, assuming that iterator found something:
+                    //  - if timestamp is null:
+                    //      - this seek will find the newest version of the next row in iterator. Exactly what we need.
+                    //  - if timestamp is not null:
+                    //      - suppose that seek key buffer has the following value: "| P0 | R0 | T0 |" (partition, row id, timestamp)
+                    //        and iterator finds something, let's denote it as "| P0 | R1 | T1 |" (partition must match). Again, there are
+                    //        few possibilities here:
+                    //          - R1 == R0, this means a match. By the rules of ordering we derive that T1 >= T0. Timestamps are stored in
+                    //            descending order, this means that we found exactly what's needed.
+                    //          - R1 > R0, this means that we found next row and T1 is either missing (pending row) or represents the latest
+                    //            version of the row. It doesn't matter in this case, because this row id will be reused to find its value
+                    //            at time T0. Additional "seek" will be required to do it.
+                    it.seek(seekKeyBuf.array());
+
+                    // Finish scan if nothing was found.
+                    if (invalid(it)) {
+                        return false;
+                    }
+
+                    // Read the actual key into a direct buffer.
+                    it.key(directBuffer);
+
+                    // To understand this condition please read that huge comment above.
+                    if (timestamp == null || !directBuffer.equals(seekKeyBufSliceWithoutTimestamp)) {
```

Comment:
The structure now is

```
if (timestamp == null || COND) {
    ACTION1
    if (timestamp != null) {
        ACTION2
    }
}
```

I suggest to transform it to

```
if (timestamp == null || COND) {
    ACTION1
}
if (timestamp != null) {
    ACTION2
}
```

This seems (arguably) easier to comprehend.

Further, `COND` can be extracted to a boolean variable with a clarifying name like `didNotFindExactRowId` (if I understood it correctly) to facilitate comprehension even more.

Reply:
I don't think that you suggest equivalent change. "timestamp != null" is not enough for ACTION2. I believe that two consecutive if-s will be more complicated, especially with the necessity to introduce a variable, which value only makes sense for the first "if"

Reply:
My bad. I meant `if (COND)` in the second `if`. In this case the variable is used in both `if`s (so twice), and also it adds some explanation to its own meaning.

To understand two embedded ifs, you need to think about them at the same time. To understand two consequtive ifs, you may think about them in sequence, independently, so consecutive ifs are usually easier to comprehend. In this case, I had to mentally apply the suggested transformation to the 'plain' structure to actually understand what is going on in this piece of code.

Reply:
Ok, I give up, let's see if it gets better